src/commands.ts

```diff
@@ -148,11 +148,79 @@ export class Commands
         this._context.outputChannel.show();
     }
 
+    packagePickerChoices(): Promise<{ label, description, pkg }[]>
+    {
+        let choices: { label, description, pkg }[] = [];
+        function addChoice(aPkg: autoproj.Package, ws: autoproj.WorkspaceInfo)
+        {
+            const choice = {
+                label: aPkg.name,
+                description: basename(ws.path),
+                pkg: aPkg
+            }
+            choices.push(choice);
+        }
+
+        return new Promise<{ label, description, pkg }[]>((resolve, reject) => {
+            let pending: any[] = [];
+            this._context.workspaces.forEachWorkspace((ws) => {
+                pending.push(ws.info().then((wsInfo) => {
+                        for (const aPkg of wsInfo.packages) {
+                            if (!this._vscode.workspaceFolders
+                                || !this._vscode.workspaceFolders.find((item) => item.uri.fsPath == aPkg[1].srcdir))
+                            {
+                                addChoice(aPkg[1], wsInfo);
+                            }
+                        }
+                    },
+                    (err) => {
+                        reject(new Error(`Could not load installation manifest: ${err.message}`));
+                    }));
+            });
+            Promise.all(pending).then(_ => {
+                choices.sort((a, b) => {
```

Comment:
Using a side variable like this is IMO much harder to follow. You could make each promise return wsInfo and the list of packages, and then aggregate then in the all(). This would actually remove the need for the addChoice side function as well.

Actually, it would even be easier to simply aggregate the wsInfo without any processing, iterate over them in the `all(...)` and sort them at the end, instead of splitting the processing into two separate promises like this.

Reply:
The point of using the two promises here is because I don't have to wait for all the `WorkspaceInfo` promises to be resolved before I start processing the packages. Granted, this might be a case of premature optimization given that I don't expect the user to be working on too many workspaces.

If we assume that, we can ditch one of the promises and use async + await instead, which will in my opinion result in the cleanest code. Thoughts?

Reply:
There is a need to use two stages of promises anyways, since you really want the loading to be done in parallel. That was not my point. My point is that there are many levels of indirection between the two, while you could do

~~~
let pending: any[] = [];
this._context.workspaces.forEachWorkspace((ws) => { pending.push(ws.info()) })
Promise.all(pending, (...wsInfo) => {
   // Do all the processing here
})
~~~

> The point of using the two promises here is because I don't have to wait for all the WorkspaceInfo promises to be resolved before I start processing the packages. Granted, this might be a case of premature optimization given that I don't expect the user to be working on too many workspaces.

Given what "work" you do on each workspace, it's not premature optimization, I think it's no optimization at all. The costly operation here is the linear search in `workspaceFolders`, which can easily be replaced by a O(1) operation with the precalculation of a Map. The rest is creating one object per package in the workspaces, which one would start to notice only if there was a few tens of thousand packages (or I'm guessing even more than that).

As for async/await vs. promises, pick the style that you prefer.

Reply:
```
let pending: any[] = [];
this._context.workspaces.forEachWorkspace((ws) => { pending.push(ws.info()) })
Promise.all(pending, (...wsInfo) => {
   // Do all the processing here
})
```

This is exactly what I was trying to avoid when I came up with the current design. By doing what you are suggesting, you are waiting for all `ws.info()` to be loaded before you start processing. `Promise.all()` won't resolve until all promises are resolved.. I wanted to have each workspaces' packages to be processed regardless of the state of the other workspaces' promises.

Reply:
Yes, I got that. But you skim probably 2ms, with a huge cost in readability. There are two indirections in this function alone: the separate helper function and the choices array.

Accounting for the change from a linear to O(1) search, for each package, you would do one Map lookup and then create one small object. We're really talking about at most a bunch of ms for thousand of packages.

Reply:
Agreed. Just wanted to make it clear. I have a fix already, will push that.